File: 01/cpp/sonar.cpp
#include <iostream>
#include <fstream>
#include <vector>

int part1(const std::vector<int>&);
int part2(const std::vector<int>&);

int main()
{
	std::ifstream infile;
	infile.open("input.txt");
	std::vector<int> input{};
	int i;

	while (infile >> i)
	{
		input.push_back(i);
	}
	infile.close();

	std::cout << "Part 1 is " << part1(input) << std::endl;
	std::cout << "Part 2 is " << part2(input) << std::endl;
	return 0;
}

int part1(const std::vector<int>& readings)
{
	int increases = 0;
	int last_depth = 32767;
	for (const int reading : readings)
	{
		if (reading > last_depth)
		{
			increases++;
		}
		last_depth = reading;
	}
	return increases;
}

int part2(const std::vector<int>& readings)
{
	int increases = 0;
	int last_depth = 32767;

	for (unsigned long x = 0; x < readings.size() - 2; x++)
	{
		const int window1 = readings[x] + readings[x + 1] + readings[x + 2];
		if (window1 > last_depth)
		{
			increases++;
		}
		last_depth = window1;
	}
	return increases;
}

File: 01/python/sonar.py
import os
import sys


def part1(readings):
    increases = 0
    last_depth = sys.maxsize

    for reading in readings:
        if reading > last_depth:
            increases = increases + 1
        last_depth = reading
    return increases


def part2(readings):
    increases = 0
    last_depth = sys.maxsize

    for pointer in range(len(readings) - 2):
        window1 = readings[pointer] + \
            readings[pointer + 1] + readings[pointer + 2]
        if window1 > last_depth:
            increases = increases + 1
        last_depth = window1
    return increases


def main():
    file_path = os.path.join(os.path.dirname(__file__), "input.txt")
    with open(file_path) as f:
        readings = f.readlines()
    readings = [int(reading) for reading in readings]
    print("Part 1 %d" % part1(readings))
    print("Part 2 %d" % part2(readings))


if __name__ == "__main__":
    sys.exit(int(main() or 0))
